sphinx.go

```diff
@@ -762,3 +811,100 @@ func (r *Router) generateSharedSecret(dhKey *btcec.PublicKey) ([sha256.Size]byte
 }
+
+// Tx is a transaction consisting of a number of sphinx packets to be atomically
+// written to the replay log. This structure helps to coordinate construction of
+// the underlying Batch object, and to ensure that the result of the processing
+// is idempotent.
+type Tx struct {
+	// batch is the set of packets to be incrementally processed and
+	// ultimately committed in this transaction
+	batch *Batch
+
+	// router is a reference to the sphinx router that created this
+	// transaction. Committing this transaction will utilize this router's
+	// replay log.
+	router *Router
+
+	// packets contains a potentially sparse list of optimistically processed
+	// packets for this batch. The contents of a particular index should
+	// only be accessed if the index is *not* included in the replay set, or
+	// otherwise failed any other stage of the processing.
+	packets []ProcessedPacket
+}
+
+// BeginTxn creates a new transaction that can later be committed back to the
+// sphinx router's replay log.
+//
+// NOTE: The nels parameter should represent the maximum number of that could be
+// added to the batch, using sequence numbers that match or exceed this value
+// could result in an out-of-bounds panic.
+func (r *Router) BeginTxn(id []byte, nels int) *Tx {
+	return &Tx{
+		batch:   NewBatch(id),
+		router:  r,
+		packets: make([]ProcessedPacket, nels),
+	}
+}
+
+// ProcessOnionPacket processes an incoming onion packet which has been forward
+// to the target Sphinx router. If the encoded ephemeral key isn't on the
+// target Elliptic Curve, then the packet is rejected. Similarly, if the
+// derived shared secret has been seen before the packet is rejected.  Finally
+// if the MAC doesn't check the packet is again rejected.
+//
+// In the case of a successful packet processing, and ProcessedPacket struct is
+// returned which houses the newly parsed packet, along with instructions on
+// what to do next.
+func (t *Tx) ProcessOnionPacket(seqNum uint16, onionPkt *OnionPacket,
+	assocData []byte) error {
+
+	// Compute the shared secret for this onion packet.
+	sharedSecret, err := t.router.generateSharedSecret(
+		onionPkt.EphemeralKey)
+	if err != nil {
+		return err
+	}
+
+	// Additionally, compute the hash prefix of the shared secret, which
+	// will serve as an identifier for detecting replayed packets.
+	hashPrefix := hashSharedSecret(&sharedSecret)
+
+	// Continue to optimistically process this packet, deferring replay
+	// protection until the end to reduce the penalty of multiple IO
+	// operations.
+	packet, err := processOnionPacket(onionPkt, &sharedSecret, assocData)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the packet's forwarding instructions, which will contain the
+	// CLTV at which the packet expires.
+	hopData := packet.ForwardingInstructions
+
+	// Add the hash prefix to pending batch of shared secrets that will be
+	// written later via Commit().
+	err = t.batch.Put(seqNum, &hashPrefix, hopData.OutgoingCltv)
```

Comment:
Similar comment here about using the CLTV of the _incoming_ HTLC, not the outgoing one.